nemo/collections/tts/models/vits.py

```diff
@@ -0,0 +1,425 @@
+            upsample_initial_channel = cfg.generator.upsample_initial_channel,
+            upsample_kernel_sizes = cfg.generator.upsample_kernel_sizes,
+            # cfg.filter_channels // 2 + 1,
+            # cfg.segment_size // cfg.train_ds.dataset.hop_length,
+            )
+        self.net_d = MultiPeriodDiscriminator(cfg.use_spectral_norm)
+        self.automatic_optimization = False
+
+        window_fn = {
+            'hann': torch.hann_window,
+            'hamming': torch.hamming_window,
+            'blackman': torch.blackman_window,
+            'bartlett': torch.bartlett_window,
+            'none': None,
+        }.get(self.hann_window, None)
+
+        self.stft = lambda x: torch.stft(
+            input=x,
+            n_fft=self.n_fft,
+            hop_length=self.hop_size,
+            win_length=self.win_length,
+            window=window_fn(self.win_length, periodic=False).to(torch.float) if window_fn else None,
+        )
+
+    def parse(self, str_input: str) -> torch.tensor:
+        # TODO: Implement
+        pass
+
+    def configure_optimizers(self):
+        self.optim_g = torch.optim.AdamW(
+            self.net_g.parameters(),
+            self._cfg.lr,
+            betas=self._cfg.betas,
+            eps=self._cfg.eps)
+        self.optim_d = torch.optim.AdamW(
+            self.net_d.parameters(),
+            self._cfg.lr,
+            betas=self._cfg.betas,
+            eps=self._cfg.eps)
+
+        scheduler_g = torch.optim.lr_scheduler.ExponentialLR(self.optim_g, gamma=self._cfg.lr_decay)
+        scheduler_g_dict = {
+            'scheduler': scheduler_g,
+            'interval': 'step',
+        }
+        scheduler_d = torch.optim.lr_scheduler.ExponentialLR(self.optim_d, gamma=self._cfg.lr_decay)
+        scheduler_d_dict = {
+            'scheduler': scheduler_d,
+            'interval': 'step'
+        }
+        return [self.optim_g, self.optim_d], [scheduler_g_dict, scheduler_d_dict]
+
+    def forward(self, batch, batch_idx):
+        with torch.no_grad():
+            # TODO: Fix
+            (x, x_lengths, spec, spec_lengths, y, y_lengths) = batch
+
+            # remove else
+            x = x[:1]
+            x_lengths = x_lengths[:1]
+
+            y_hat, attn, mask, *_ = self.net_g.module.infer(x, x_lengths, max_len=1000)
+            y_hat_lengths = mask.sum([1, 2]).long() * self._cfg.hop_size
+
+        return y_hat, y_lengths
+
+
+    def get_spec(self, audio):
+        with torch.cuda.amp.autocast(enabled=False):
+            spec = self.stft(audio)
+            if spec.dtype in [torch.cfloat, torch.cdouble]:
+                spec = torch.view_as_real(spec)
+            spec = torch.sqrt(spec.pow(2).sum(-1) + 1e-9)
+        return spec
+
+    def training_step(self, batch, batch_idx):
+        # (x, x_lengths, y, y_lengths) = batch
+        (y, y_lengths, x, x_lengths) = batch
+
+        spec = self.get_spec(y)
+        spec_lengths = torch.ones(spec.shape[0]) * spec.shape[2]
```

Comment:
As I mentioned in the meeting, you need to get accurate spec_lengths

Reply:
done.

Reply:
done.